controller/service.go

```diff
@@ -260,3 +266,18 @@ func isEqualIPs(ipsA, ipsB []net.IP) bool {
 	})
 	return reflect.DeepEqual(ipsA, ipsB)
 }
+
+// updateSvcStatus updating svc's status and svcRo's status
+func updateSvcStatus(svc, svcRo *v1.Service) {
+	var st v1.ServiceStatus
+	st, svc = svc.Status, svcRo.DeepCopy()
+	// we need be update svcRo's status
```

Comment:
Why do we need this? svcRo is supposed to be read only

Reply:
We should update status and annotations an same time , and then call `UpdateService`, Update svcRo's status here, I think it's OK.

Reply:
Otherwise, when `updateSvcAnnotations ` is called, the parameter `svc` has been reassigned and annotations are lost

Reply:
uhm, never easy... let me think about it, not sure I like this versin because we'll need to remember if we ever add another function that changes something else

Reply:
So (and sorry if I stress too much on this), I think we can keep the "defensive" spirit of the original code by doing this:

```go
	if reflect.DeepEqual(svcRo, svc) {
		level.Debug(l).Log("event", "noChange", "msg", "service converged, no change")
		return successRes
	}

	toWrite := svcRo.DeepCopy()
	if !reflect.DeepEqual(toWrite.Status, svc.Status) {
		toWrite.Status = svc.Status
	}

	if !reflect.DeepEqual(toWrite.Annotations, svc.Annotations) {
		toWrite.Annotatins = svc.Annotations
	}

	if err := c.client.UpdateStatus(svc); err != nil {
		level.Error(l).Log("op", "updateServiceStatus", "error", err, "msg", "failed to update service")
		return controllers.SyncStateError
	}
```

In theory, the only two things convergeBalancer is supposed to change are the annotations and the status.
We can be even more cautious and call updateStatus when at least one of the two changes

Reply:
No sorry, Your suggestions make sense.

Agree with you, and do we also need a flag to decide `UpdateStatus`? When annotations or status changed.